### PR TITLE
added a fix for reference counting with python callbacks

### PR DESCRIPTION
**Pull request by jordan9001**
PR created in response to https://github.com/JonathanSalwan/Triton/issues/1035

This has more changes than the patch referenced in that issue. This causes Triton to now throw an exception if you try to remove a callback and do not find the callback. If the same callback is registered multiple times, previously one call to remove would remove all the duplicates as well. With this update remove must be called the same number of times as add.

This was added because it is necessary to prevent bad dereferences on python objects passed to `removeCallback`.

### CODE REVIEW

File: src/libtriton/arch/x86/x86Specifications.cpp [hunk 1]
```diff
@@ -50,7 +50,7 @@ namespace triton {
                                                       X86_LOWER,                                            \
                                                       true)                                                 \
                               );                                                                            \
-            name2id.emplace(#LOWER_NAME, ID_REG_X86_##UPPER_NAME);
+          name2id.emplace(#LOWER_NAME, ID_REG_X86_##UPPER_NAME);
```
Comment (jordan9001) [inline src/libtriton/arch/x86/x86Specifications.cpp:53]: This was done to remove compile time warnings. This doesn't change the way the code works, but if this _was_ intended to be part of the above if statement, then brackets need to be added.

**Comment by JonathanSalwan** [comment]
Thanks for this MR.

Looks like there is an unittest that does not pass (src/testers/unittests/test_callback.py). This unittest tests if there is a memory leak by counting Python references. Maybe we added to much `Py_INCREF` or miss one `Py_DECCREF`.

```python
def method_callback(self, api, _):                                          
     pass

def test_method_callback_removal(self):                                     
    # regression test for #809                                              
    import sys                                                              
    self.Triton = TritonContext()                                           
    self.Triton.setArchitecture(ARCH.X86_64)                                
    cb = self.method_callback.__func__                                      
    cb_initial_refcnt = sys.getrefcount(cb)                                 
    self.Triton.addCallback(CALLBACK.GET_CONCRETE_MEMORY_VALUE, self.method_callback)
    self.Triton.removeCallback(CALLBACK.GET_CONCRETE_MEMORY_VALUE, self.method_callback)
    self.assertTrue(cb_initial_refcnt == sys.getrefcount(cb))               
```

**Comment by jordan9001** [comment]
With the updates I think it is correct. is there anything else you need from me @JonathanSalwan ?

**Comment by JonathanSalwan** [comment]
Looks ok for me. Thanks @jordan9001! I will merge it this week :)

Cheers,

**Pull request merged**